In [1]:
import pandas as pd
import pickle
import preprocess_utils
import numpy as np
import json

from_database = False

if from_database:
    # Load tables from database 
    dfs = preprocess_utils.load_all_tables()
else:
    # Open the file in binary read mode and load the pickle data
    with open('preprocessed_data/dfs.pickle', 'rb') as f:
        dfs = pickle.load(f)
        for key, df in dfs.items():
            print(f"DataFrame {key:<20} has {len(df):<6} rows and {len(df.columns):<6} columns.")

if from_database:
    # Save tables to pickle
    with open('preprocessed_data/dfs.pickle', 'wb') as f:
        pickle.dump(dfs, f)

    # load it back with:
    with open('preprocessed_data/dfs.pickle', 'rb') as f:
        dfs = pickle.load(f)

DataFrame abuse_report_types   has 4      rows and 4      columns.
DataFrame abuse_reports        has 0      rows and 8      columns.
DataFrame ar_internal_metadata has 1      rows and 4      columns.
DataFrame articles             has 400    rows and 33     columns.
DataFrame assertion_types      has 10     rows and 11     columns.
DataFrame assertion_versions   has 19668  rows and 10     columns.
DataFrame assertions           has 13941  rows and 14     columns.
DataFrame assertions_genes     has 2053   rows and 2      columns.
DataFrame assertions_tags      has 403    rows and 2      columns.
DataFrame assertions_technique_types has 0      rows and 2      columns.
DataFrame assertions_techniques has 0      rows and 2      columns.
DataFrame assessment_types     has 13     rows and 9      columns.
DataFrame assessments          has 0      rows and 6      columns.
DataFrame authors              has 797    rows and 10     columns.
DataFrame banned_orcid_users   has 0      rows and 5   

## A. Database articles, claims and authors

### A.1. Process journal articles from the database

In [2]:
# Process articles
articles = preprocess_utils.clean_df_from_database(dfs["articles"])
articles = articles.rename(columns={"id": "article_id"})
# add journals to articles:
journals = preprocess_utils.clean_df_from_database(dfs["journals"])
journals = journals.drop('tag', axis=1).rename(columns={"id": "journal_id", "name": "journal_name"})
articles = articles.merge(journals, on="journal_id", how="left", suffixes=('', '_journal')).drop("journal_id", axis=1)


# Extract first and last authors from authors_txt
def extract_authors(authors_txt):
    if isinstance(authors_txt, str) and authors_txt:
        authors_list = authors_txt.split(';')
        first_author = authors_list[0].strip() if authors_list else None
        last_author = authors_list[-1].strip() if authors_list else None
        return pd.Series([first_author, last_author])
    return pd.Series([None, None])

# Apply the function to create new columns
articles[['first_author_extracted', 'last_author_extracted']] = articles['authors_txt'].apply(extract_authors)

# Check the results
print(f"Total articles: {len(articles)}")
print(f"Articles with first author: {articles['first_author_extracted'].notna().sum()}")
print(f"Articles with last author: {articles['last_author_extracted'].notna().sum()}")

articles = articles.drop(columns=["key", "references_txt", "additional_context", "num",  "nber_panels", "large_scale", "nber_tables", "published_at"])


# Display the first few rows to verify
articles[['authors_txt', 'first_author_extracted', 'last_author_extracted']].head()

Total articles: 400
Articles with first author: 400
Articles with last author: 400


authors_txt first_author_extracted  \
0                                  Rizki RM;Rizki TM               Rizki RM   
1             Leulier F;Lhocine N;Lemaitre B;Meier P              Leulier F   
2  Kim YS;Han SJ;Ryu JH;Choi KH;Hong YS;Chung YH;...                 Kim YS   
3                      Tingvall TO;Roos E;Engström Y            Tingvall TO   
4                   Bhaskar V;Valentine SA;Courey AJ              Bhaskar V   

  last_author_extracted  
0              Rizki TM  
1               Meier P  
2                Lee WJ  
3            Engström Y  
4             Courey AJ

In [3]:


# We need to analyze and clean up affiliations in the articles dataframe

def extract_primary_affiliation(affs_json_str):
    """Extract the first affiliation from the JSON string, safely handle bad JSON format."""
    try:
        # Parse the JSON string to a Python object
        affs_list = json.loads(affs_json_str)
        
        # Get the first non-empty affiliation
        for aff in affs_list:
            if aff and isinstance(aff, list) and len(aff) > 0 and aff[0].strip():
                return aff[0].strip()
        
        return None  # No valid affiliation found
    except (json.JSONDecodeError, TypeError, IndexError):
        # Handle poorly formatted JSON or other errors
        return None

# Create a new column with clean affiliations
articles['primary_affiliation'] = articles['affs_json'].apply(extract_primary_affiliation)

# Count articles with multiple vs. single affiliations
def count_affiliations(affs_json_str):
    """Count the number of non-empty affiliations in the JSON string."""
    try:
        affs_list = json.loads(affs_json_str)
        count = sum(1 for aff in affs_list if aff and isinstance(aff, list) and len(aff) > 0 and aff[0].strip())
        return count
    except (json.JSONDecodeError, TypeError, IndexError):
        return 0

articles['affiliation_count'] = articles['affs_json'].apply(count_affiliations)

# Identify articles with multiple affiliations
multiple_aff_articles = articles[articles['affiliation_count'] > 1]

# Print summary statistics
print(f"Total articles: {len(articles)}")
print(f"Articles with 1 affiliation: {len(articles[articles['affiliation_count'] == 1])}")
print(f"Articles with multiple affiliations: {len(multiple_aff_articles)}")
print(f"Articles with no valid affiliations: {len(articles[articles['affiliation_count'] == 0])}")

# Sample of articles with multiple affiliations
if not multiple_aff_articles.empty:
    print("\nSample of articles with multiple affiliations:")
    sample = multiple_aff_articles.head(5)
    for _, row in sample.iterrows():
        print(f"Article ID: {row['article_id']}, Title: {row['title'][:50]}...")
        try:
            affs = json.loads(row['affs_json'])
            for i, aff in enumerate(affs):
                if aff and isinstance(aff, list) and len(aff) > 0 and aff[0].strip():
                    print(f"  Affiliation {i+1}: {aff[0]}")
        except:
            print("  Error parsing affiliations")
        print()

Total articles: 400
Articles with 1 affiliation: 387
Articles with multiple affiliations: 0
Articles with no valid affiliations: 13


In [4]:
articles = articles.drop(columns=["affs_json", "affiliation_count"])

# Update impact factor for "Proceedings. Biological sciences" and standardize the journal name
articles.loc[articles["journal_name"] == "Proceedings. Biological sciences", "impact_factor"] = 4.7 
articles.loc[articles["journal_name"] == "Proceedings. Biological sciences", "journal_name"] = "Proceedings Biological Sciences"


In [5]:
articles = articles.apply(preprocess_utils.safe_strip)
articles.to_csv("preprocessed_data/articles_db.csv", index=False)

In [6]:
articles

article_id                                        authors_txt  \
0          3347                                  Rizki RM;Rizki TM   
1          1987             Leulier F;Lhocine N;Lemaitre B;Meier P   
2          2761  Kim YS;Han SJ;Ryu JH;Choi KH;Hong YS;Chung YH;...   
3          2649                      Tingvall TO;Roos E;Engström Y   
4          3441                   Bhaskar V;Valentine SA;Courey AJ   
..          ...                                                ...   
395        1882        Derré I;Pypaert M;Dautry-Varsat A;Agaisse H   
396        1815    Schmidt RL;Trejo TR;Plummer TB;Platt JL;Tang AH   
397        1884     Avet-Rochex A;Perrin J;Bergeret E;Fauvarque MO   
398        2041                Tanji T;Ohashi-Kobayashi A;Natori S   
399        1926  Kurucz E;Márkus R;Zsámboki J;Folkl-Medzihradsz...   

                                                 title      pmid volume issue  \
0    Cell interactions in the differentiation of a ...    111992     12     3   
1    The Drosophila inhibitor of apoptosis protein ...  16894030     26    21   
2    Lipopolysaccharide-activated kinase, an essent...  10636911    275     3   
3    The GATA factor Serpent is required for the on...  11274409     98     7   
4    A functional interaction between dorsal and co...  10660560    275     6   
..                                                 ...       ...    ...   ...   
395  RNAi screen in Drosophila cells reveals the in...  17967059      3    10   
396  Infection-induced proteolysis of PGRP-LC contr...  18308747     22     3   
397  Rac2 is a major actor of Drosophila resistance...  17903178     12    10   
398  Participation of a galactose-specific C-type l...  16475980    396     1   
399  Nimrod, a putative phagocytosis receptor with ...  17363253     17     7   

                                              abstract  year  \
0    The cellular events in the formation of melano...  1979   
1    The founding member of the inhibitor of apopto...  2006   
2    Eukaryotic organisms use a similar Rel/NF-kapp...  2000   
3    Innate immunity in Drosophila is characterized...  2001   
4    To identify proteins that regulate the functio...  2000   
..                                                 ...   ...   
395  Chlamydia spp. are intracellular obligate bact...  2007   
396  The Drosophila immune deficiency (IMD) pathway...  2008   
397  Pathogen recognition and engulfment by phagocy...  2007   
398  A galactose-specific C-type lectin has been pu...  2006   
399  The hemocytes, the blood cells of Drosophila, ...  2007   

                                          journal_name  impact_factor  \
0    Differentiation; research in biological diversity           3.39   
1                       Molecular and cellular biology           5.30   
2                  The Journal of biological chemistry           4.80   
3    Proceedings of the National Academy of Science...          11.10   
4                  The Journal of biological chemistry           4.80   
..                                                 ...            ...   
395                                     PLoS pathogens           6.70   
396  FASEB journal : official publication of the Fe...           4.80   
397  Genes to cells : devoted to molecular & cellul...           2.10   
398                            The Biochemical journal           4.10   
399                               Current biology : CB           9.20   

    first_author_extracted last_author_extracted  \
0                 Rizki RM              Rizki TM   
1                Leulier F               Meier P   
2                   Kim YS                Lee WJ   
3              Tingvall TO            Engström Y   
4                Bhaskar V             Courey AJ   
..                     ...                   ...   
395                Derré I             Agaisse H   
396             Schmidt RL               Tang AH   
397          Avet-Rochex A          Fauvarque MO   
398                Tanji 

### A.2. Process claims from the database

claims is merged with articles


In [7]:
# Main processing
claims = dfs["assertions"].copy()
print(len(claims))
claims = claims[claims["obsolete"] == False].copy()
print(len(claims))
claims = preprocess_utils.clean_df_from_database(claims)

id_cols = [col for col in claims.columns if "_id" in col]
print(id_cols)

claims = claims.merge(articles, on="article_id", how="left", suffixes=('', '_article'))
id_cols = [col for col in claims.columns if "_id" in col]
print(id_cols)

# Process assertion types
assertion_types = preprocess_utils.clean_df_from_database(dfs["assertion_types"])
assertion_types = assertion_types.rename(columns={"id": "assertion_type_id", "name": "assertion_type"})
claims = claims.merge(assertion_types, on="assertion_type_id", how="left", suffixes=('', '_assertion_type')).drop("assertion_type_id", axis=1)

# Process assessment types
assessment_types = preprocess_utils.clean_df_from_database(dfs["assessment_types"])
assessment_types = assessment_types.rename(columns={"id": "assessment_type_id", "name": "assessment_type"})
claims = claims.merge(assessment_types, on="assessment_type_id", how="left", suffixes=('', '_assessment_type')).drop("assessment_type_id", axis=1)

id_cols = [col for col in claims.columns if "_id" in col]
print(id_cols)

13941
13299
['article_id', 'assertion_type_id', 'assessment_type_id']
['article_id', 'assertion_type_id', 'assessment_type_id']
['article_id']


In [8]:
claims = claims.drop(['badge_tag_classes','description'], axis=1) # most not consistently used accross dataset
claims = claims.set_index('id')

In [9]:
claims = claims.apply(preprocess_utils.safe_strip)
claims.to_csv('preprocessed_data/claims_db.csv')

string_columns = claims.select_dtypes(include='object').columns.drop(["assessment_type", "journal_name"])
claims_truncated = claims.copy()
for col in string_columns:
    if col in claims_truncated.columns:
        claims_truncated[col] = claims_truncated[col].apply(lambda x: preprocess_utils.truncate_string(x))

# Save truncated dataframe
claims_truncated.to_csv('preprocessed_data/claims_db_truncated_for_llm.csv', index=False)
claims_truncated

content  article_id  rank  pmid  \
id                                                       
5838   <p>Belozerov VE, Lin...      2049.0     1  None   
693                                 2517.0     1  None   
695                                 2517.0     1  None   
5840   Verified by many sub...      2204.0     1  None   
5839   Activation of a cell...      2204.0     1  None   
...                        ...         ...   ...   ...   
14468  The DIF antiserum us...      2898.0     1  None   
11330  Although independent...      1577.0     1  None   
11329  Spatzle (#gene:FBgn0...      1577.0     1  None   
11350  Although independent...      1577.0     1  None   
11349  Spatzle (#gene:FBgn0...      1577.0     3  None   

                   authors_txt                    title  pmid_article volume  \
id                                                                             
5838   Zhuang ZH;Zhou Y;Yu ...  Regulation of Drosop...    16014325.0     18   
693    Choe KM;Werner T;Stö...  Requirement for a pe...    11872802.0    296   
695    Choe KM;Werner T;Stö...  Requirement for a pe...    11872802.0    296   
5840   Zettervall CJ;Anderl...  A directed screen fo...    15381778.0    101   
5839   Zettervall CJ;Anderl...  A directed screen fo...    15381778.0    101   
...                        ...                      ...           ...    ...   
14468  Williams MJ;Rodrigue...  The 18-wheeler mutat...     9321392.0     16   
11330  Shia AK;Glittenberg ...  Toll-dependent antim...    19934223.0    122   
11329  Shia AK;Glittenberg ...  Toll-dependent antim...    19934223.0    122   
11350  Shia AK;Glittenberg ...  Toll-dependent antim...    19934223.0    122   
11349  Shia AK;Glittenberg ...  Toll-dependent antim...    19934223.0    122   

       issue                 abstract  ...  \
id                                     ...   
5838       4  The p38 mitogen-acti...  ...   
693     5566  Components of microb...  ...   
695     5566  Components of microb...  ...   
5840      39  An attack by a paras...  ...   
5839      39  An attack by a paras...  ...   
...      ...                      ...  ...   
14468     20  Mammals and insects ...  ...   
11330  Pt 24  In Drosophila, the h...  ...   
11329  Pt 24  In Drosophila, the h...  ...   
11350  Pt 24  In Drosophila, the h...  ...   
11349  Pt 24  In Drosophila, the h...  ...   

                                            journal_name impact_factor  \
id                                                                       
5838                                 Cellular signalling           4.8   
693                             Science (New York, N.Y.)          56.9   
695                             Science (New York, N.Y.)          56.9   
5840   Proceedings of the National Academy of Science...          11.1   
5839   Proceedings of the National Academy of Science...          11.1   
...                                                  ...           ...   
14468                                   The EMBO journal          11.4   
11330                            Journal of cell science           4.0   
11329                            Journal of cell science           4.0   
11350                            Journal of cell science           4.0   
11349                            Journal of cell science           4.0   

       first_author_extracted last_author_extracted      primary_affiliation  \
id                                                                             
5838                Zhuang ZH                 Ge BX  Signal Transduction ...   
693                   Choe KM           Anderson KV  Molecular Biology Pr...   
695                   Choe KM           Anderson KV  Molecular Biology Pr...   
5840            Zettervall CJ            Hultmark D  Umeå Centre for Mole...   
5839            Zettervall CJ            Hultmark D  Umeå Centre for Mole...   
...                       ...                   ...                      ...   
14468             William

In [10]:
claims_truncated

content  article_id  rank  pmid  \
id                                                       
5838   <p>Belozerov VE, Lin...      2049.0     1  None   
693                                 2517.0     1  None   
695                                 2517.0     1  None   
5840   Verified by many sub...      2204.0     1  None   
5839   Activation of a cell...      2204.0     1  None   
...                        ...         ...   ...   ...   
14468  The DIF antiserum us...      2898.0     1  None   
11330  Although independent...      1577.0     1  None   
11329  Spatzle (#gene:FBgn0...      1577.0     1  None   
11350  Although independent...      1577.0     1  None   
11349  Spatzle (#gene:FBgn0...      1577.0     3  None   

                   authors_txt                    title  pmid_article volume  \
id                                                                             
5838   Zhuang ZH;Zhou Y;Yu ...  Regulation of Drosop...    16014325.0     18   
693    Choe KM;Werner T;Stö...  Requirement for a pe...    11872802.0    296   
695    Choe KM;Werner T;Stö...  Requirement for a pe...    11872802.0    296   
5840   Zettervall CJ;Anderl...  A directed screen fo...    15381778.0    101   
5839   Zettervall CJ;Anderl...  A directed screen fo...    15381778.0    101   
...                        ...                      ...           ...    ...   
14468  Williams MJ;Rodrigue...  The 18-wheeler mutat...     9321392.0     16   
11330  Shia AK;Glittenberg ...  Toll-dependent antim...    19934223.0    122   
11329  Shia AK;Glittenberg ...  Toll-dependent antim...    19934223.0    122   
11350  Shia AK;Glittenberg ...  Toll-dependent antim...    19934223.0    122   
11349  Shia AK;Glittenberg ...  Toll-dependent antim...    19934223.0    122   

       issue                 abstract  ...  \
id                                     ...   
5838       4  The p38 mitogen-acti...  ...   
693     5566  Components of microb...  ...   
695     5566  Components of microb...  ...   
5840      39  An attack by a paras...  ...   
5839      39  An attack by a paras...  ...   
...      ...                      ...  ...   
14468     20  Mammals and insects ...  ...   
11330  Pt 24  In Drosophila, the h...  ...   
11329  Pt 24  In Drosophila, the h...  ...   
11350  Pt 24  In Drosophila, the h...  ...   
11349  Pt 24  In Drosophila, the h...  ...   

                                            journal_name impact_factor  \
id                                                                       
5838                                 Cellular signalling           4.8   
693                             Science (New York, N.Y.)          56.9   
695                             Science (New York, N.Y.)          56.9   
5840   Proceedings of the National Academy of Science...          11.1   
5839   Proceedings of the National Academy of Science...          11.1   
...                                                  ...           ...   
14468                                   The EMBO journal          11.4   
11330                            Journal of cell science           4.0   
11329                            Journal of cell science           4.0   
11350                            Journal of cell science           4.0   
11349                            Journal of cell science           4.0   

       first_author_extracted last_author_extracted      primary_affiliation  \
id                                                                             
5838                Zhuang ZH                 Ge BX  Signal Transduction ...   
693                   Choe KM           Anderson KV  Molecular Biology Pr...   
695                   Choe KM           Anderson KV  Molecular Biology Pr...   
5840            Zettervall CJ            Hultmark D  Umeå Centre for Mole...   
5839            Zettervall CJ            Hultmark D  Umeå Centre for Mole...   
...                       ...                   ...                      ...   
14468             William

### A.3. Process authors from database

there are about two author per article... though the total of authors is 797 so some are missing.

In [11]:
a = preprocess_utils.clean_df_from_database(dfs["authors"])  
a[a["article_id"] == 2053]

id  article_id        name  sex  career_stage_id  leading_author  \
0   2        2053      Brun S    1                3           False   
1   3        2053  Lemaitre B    1                4            True   

   first_author  expertise_level_id  
0          True                   2  
1         False                   2

In [12]:
len(a)

797

In [13]:
a[a["name"] == "Lemaitre B"].head()

id  article_id        name  sex  career_stage_id  leading_author  \
1    3        2053  Lemaitre B    1                4            True   
7    8        3437  Lemaitre B    1                4            True   
14  16        2801  Lemaitre B    1                4            True   
30  36        2617  Lemaitre B    1                4            True   
44  56        2869  Lemaitre B    1                4            True   

    first_author  expertise_level_id  
1          False                   2  
7          False                   2  
14         False                   2  
30         False                   2  
44         False                   2

In [14]:
a[a["article_id"] == 2451]

id  article_id           name  sex  career_stage_id  leading_author  \
17    19        2451  De Gregorio E    1                3           False   
95    20        2451     Lemaitre B    1                5           False   
789  795        2451        Brey PT    1                5            True   

     first_author  expertise_level_id  
17           True                   2  
95          False                   2  
789         False                   2

In [15]:
authors = preprocess_utils.clean_df_from_database(dfs["authors"].copy())
authors
career_stage = dfs["career_stages"].rename(columns={"id": "career_stage_id"})
authors = authors.merge(career_stage, on="career_stage_id", how="left", suffixes=('', '_career_stage')).drop("career_stage_id", axis=1)
expertise_level = dfs["expertise_levels"].rename(columns={"id": "expertise_level_id"})
authors = authors.merge(expertise_level, on="expertise_level_id", how="left", suffixes=('', '_expertise_level')).drop("expertise_level_id", axis=1)
authors = authors.rename(columns={"name_career_stage": "career_stage", "name_expertise_level": "expertise_level", "id": "author_id"})

# Convert sex values: 1 to "Male", 0 to "Female"
authors.loc[authors["sex"] == 1, "sex"] = "Male"
authors.loc[authors["sex"] == 0, "sex"] = "Female"
authors["author_key"] =  authors["name"].str.lower().str.normalize('NFKD').str.encode('ascii', errors='ignore').str.decode('utf-8')

authors

/var/folders/pj/15hxgl0j1wg2w_t5k7bvnq5c0000gn/T/ipykernel_25753/706437459.py:10: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'Male' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  authors.loc[authors["sex"] == 1, "sex"] = "Male"


author_id  article_id               name     sex  leading_author  \
0            2        2053             Brun S    Male           False   
1            3        2053         Lemaitre B    Male            True   
2            1        1650         Charroux B    Male           False   
3           97        2149        Söderhäll K    Male            True   
4            5        2278        Silverman N    Male           False   
..         ...         ...                ...     ...             ...   
792        797        2067         Cerenius L    Male            True   
793        798        2122         Higgins DE    Male            True   
794        799        2184  Mengin-Lecreulx D    Male            True   
795        800        2130           Moore KJ  Female            True   
796        700        1990           Matova N  Female           False   

     first_author career_stage expertise_level         author_key  
0            True     Post-doc     Experienced             brun s  
1           False    Junior PI     Experienced         lemaitre b  
2            True     Post-doc     Experienced         charroux b  
3           False    Senior PI     Experienced        soderhall k  
4            True     Post-doc        Newcomer        silverman n  
..            ...          ...             ...                ...  
792         False    Senior PI     Experienced         cerenius l  
793         False    Senior PI        Newcomer         higgins de  
794         False    Senior PI     Experienced  mengin-lecreulx d  
795         False    Junior PI        Newcomer           moore kj  
796          True     Post-doc        Newcomer           matova n  

[797 rows x 9 columns]

In [16]:
# check authors thare not both leading and first authors, or that are neither using XOR
authors[~(authors["leading_author"] ^ authors["first_author"])]

author_id  article_id           name     sex  leading_author  \
95          20        2451     Lemaitre B    Male           False   
194        127        2113   Schneider DS    Male           False   
195        197        3440     Lemaitre B    Male           False   
196        169        2184     Lemaitre B    Male           False   
243        250        1947   Schneider DS    Male            True   
256        260        1763   Schneider DS    Male           False   
257        252        2730  Shahabuddin M    Male           False   
287        303        3458     Hultmark D    Male            True   
329        251        2730   Schneider DS    Male            True   
334        378        1702    Williams MJ    Male            True   
342        365        3447       Rizki TM    Male            True   
350        363        3448       Rizki TM    Male            True   
355        364        3292       Rizki TM    Male            True   
356        370        3450       Gateff E  Female            True   
368        362        3273       Rizki TM    Male            True   
371        371        3449       Gateff E  Female            True   
473        483        2764       Imler JL    Male            True   
565        497        2122     Perrimon N    Male           False   
658        647        3446       Nappi AJ    Male            True   
686        702        1762      Wegner KM    Male            True   
741        720        2067    Söderhäll K    Male           False   
742        706        2130   Ezekowitz RA    Male           False   
790        792        3273    Bellotti RA    Male           False   

     first_author career_stage expertise_level     author_key  
95          False    Senior PI     Experienced     lemaitre b  
194         False    Junior PI     Experienced   schneider ds  
195         False    Junior PI     Experienced     lemaitre b  
196         False    Junior PI     Experienced     lemaitre b  
243          True    Senior PI     Experienced   schneider ds  
256         False    Senior PI     Experienced   schneider ds  
257         False    Junior PI        Newcomer  shahabuddin m  
287          True    Senior PI     Experienced     hultmark d  
329          True    Junior PI        Newcomer   schneider ds  
334          True    Junior PI     Experienced    williams mj  
342          True    Senior PI     Experienced       rizki tm  
350          True    Senior PI     Experienced       rizki tm  
355          True    Senior PI     Experienced       rizki tm  
356          True    Senior PI     Experienced       gateff e  
368          True     Post-doc     Experienced       rizki tm  
371          True    Senior PI     Experienced       gateff e  
473          True    Junior PI        Newcomer       imler jl  
565         False    Senior PI     Experienced     perrimon n  
658          True    Junior PI     Experienced       nappi aj  
686          True     Post-doc        Newcomer      wegner km  
741         False    Senior PI     Experienced    soderhall k  
742         False    Senior PI        Newcomer   ezekowitz ra  
790         False       Master        Newcomer    bellotti ra

In [17]:
#authors = authors.merge(articles, on="article_id", how="left", suffixes=('', '_article'))
#authors.sort_values(by=["article_id", "author_id", ])

authors = preprocess_utils.clean_author_keys(authors, "author_key")
authors = authors.apply(preprocess_utils.safe_strip)
authors.to_csv('preprocessed_data/authors_db.csv', index=False)

In [18]:
authors.sort_values(by=["author_key"])

author_id  article_id          name     sex  leading_author  \
724        745        1535  Abdelsadik A    Male           False   
485        496        2122     Agaisse H    Male           False   
487        499        1882     Agaisse H    Male            True   
15          17        2297     Agaisse H    Male           False   
104        108        1752    Aggarwal K  Female           False   
..         ...         ...           ...     ...             ...   
754        784        1491         Zhu S    Male            True   
740        714        1756         Zhu S    Male            True   
653        621        1748         Zhu S    Male            True   
588        588        2025     Zhuang ZH    Male           False   
51          68        2049     Zhuang ZH    Male           False   

     first_author career_stage expertise_level    author_key  
724          True          PhD     Experienced  abdelsadik a  
485          True     Post-doc     Experienced     agaisse h  
487         False    Junior PI     Experienced     agaisse h  
15           True     Post-doc     Experienced     agaisse h  
104          True          PhD     Experienced    aggarwal k  
..            ...          ...             ...           ...  
754         False    Junior PI        Newcomer         zhu s  
740         False    Junior PI        Newcomer         zhu s  
653         False    Senior PI        Newcomer         zhu s  
588          True          PhD        Newcomer     zhuang zh  
51           True          PhD        Newcomer     zhuang zh  

[797 rows x 9 columns]

## B. Process from the XLSX files

- **3:citations** `input_data/2025-03-28/2025-03-27_dropbox_6_ReproSci/0_last version/2025_Last version_March 19th/3_2025_March 9th_citation_counts.xlsm`
  -  il s’agit des citations : je ne l’ai pas travaillé.  
 - **4:first&last** `input_data/2025-03-28/2025-03-27_dropbox_6_ReproSci/0_last version/2025_Last version_March 19th/4_working first and last.xlsx`
    -   ils’agit d’une liste de 20 auteurs qui sont first and last and mais ne pas prendre en compte ceux qui sont PI.  

 It seems that
 - sex -> FH
 - PhD Post-doc -> FH
 - Become a Pi -> FH
 - current job -> FH
 - MD -> FH
 - Affiliation -> claim
 - Country -> claim
 - Ivy League -> claim

### B.1. Read the Triage last_author files which contains all pairs

update April 7, 2025: we don't care avout this file anymore

In [ ]:
# Read Author info, which contains all the pairs
paper_auth_pairs = pd.read_excel('input_data/2025-02-14_last_xlsx/1_Triage_Last author.xlsx', sheet_name='Tri sans les doublons')
# Drop all columns with 'Unnamed' in the name
paper_auth_pairs = paper_auth_pairs.drop(columns=[col for col in paper_auth_pairs.columns if 'Unnamed' in col]).drop(columns=['Source'])
# rename RIZKI MT	to Rizki TM in column last author
paper_auth_pairs['last author'] = paper_auth_pairs['last author'].replace({'Rizki MT': 'Rizki TM','RIZKI MT': 'Rizki TM' })

paper_auth_pairs.loc[paper_auth_pairs["MD"] == "??", "MD"] = np.nan
paper_auth_pairs.loc[paper_auth_pairs["MD"] == "?", "MD"] = np.nan
paper_auth_pairs.loc[paper_auth_pairs["MD"] == 'O', "MD"] = False
paper_auth_pairs.loc[paper_auth_pairs["MD"] == 0, "MD"] = False
paper_auth_pairs.loc[paper_auth_pairs["MD"] == 1, "MD"] = True

paper_auth_pairs.loc[paper_auth_pairs["Become a Pi"] == "??", "Become a Pi"] = np.nan
paper_auth_pairs.loc[paper_auth_pairs["Become a Pi"] == 0, "Become a Pi"] = False
paper_auth_pairs.loc[paper_auth_pairs["Become a Pi"] == 1, "Become a Pi"] = True

paper_auth_pairs.loc[paper_auth_pairs["Ivy league"] == 0, "Ivy league"] = False
paper_auth_pairs.loc[paper_auth_pairs["Ivy league"] == 1, "Ivy league"] = True

paper_auth_pairs.loc[paper_auth_pairs["Sex"] == "??", "Sex"] = np.nan
paper_auth_pairs.loc[paper_auth_pairs["Sex"] == 1, "Sex"] = "Male"
paper_auth_pairs.loc[paper_auth_pairs["Sex"] == 0, "Sex"] = "Female"

paper_auth_pairs.loc[paper_auth_pairs["current job"] == "pi", "current job"] = "PI"
paper_auth_pairs.loc[paper_auth_pairs["current job"] == "Pi", "current job"] = "PI"
paper_auth_pairs.loc[paper_auth_pairs["current job"] == "?", "current job"] = np.nan
paper_auth_pairs.loc[paper_auth_pairs["current job"] == "??", "current job"] = np.nan

# create merge columns: lowercased and stripped of accents
paper_auth_pairs['first_author_key'] = paper_auth_pairs['first author'].str.lower()
paper_auth_pairs['first_author_key'] = paper_auth_pairs['first_author_key'].str.normalize('NFKD').str.encode('ascii', errors='ignore').str.decode('utf-8')
paper_auth_pairs = preprocess_utils.clean_author_keys(paper_auth_pairs, "first_author_key")

paper_auth_pairs.loc[:, 'last_author_key'] = paper_auth_pairs['last author'].str.lower()
paper_auth_pairs.loc[:, 'last_author_key'] = (paper_auth_pairs['last_author_key']
    .str.normalize('NFKD')
    .str.encode('ascii', errors='ignore')
    .str.decode('utf-8'))
#paper_auth_pairs.loc[:, 'last_author_key'] = paper_auth_pairs['last_author_key'].str.replace('rizki mt', 'rizki tm')
paper_auth_pairs = preprocess_utils.clean_author_keys(paper_auth_pairs, "last_author_key")


print(paper_auth_pairs["Sex"].unique(), paper_auth_pairs["MD"].unique(), paper_auth_pairs["Become a Pi"].unique(), paper_auth_pairs["Ivy league"].unique())

paper_auth_pairs = paper_auth_pairs.apply(preprocess_utils.safe_strip())


['Female' 'Male' nan] [False nan True] [True False nan] [True False]


/var/folders/pj/15hxgl0j1wg2w_t5k7bvnq5c0000gn/T/ipykernel_25753/687267990.py:18: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'False' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  paper_auth_pairs.loc[paper_auth_pairs["Ivy league"] == 0, "Ivy league"] = False


#### Merge it with database to obtain the article ID

In [55]:
paper_auth_pairs

last author first author     Sex PhD Post-doc Become a Pi  \
0      Agaisse H        Derré  Female     Post-doc        True   
1    Aguilera RJ     Seong CS    Male          PhD       False   
2       Aigaki T      Tsuda M    Male          PhD        True   
3           Ando       Markus    Male          PhD       False   
4           Ando        Rus F  Female          PhD       False   
..           ...          ...     ...          ...         ...   
315        Yu XQ         Ao J     NaN          PhD         NaN   
316        Zhu S       Yuan Y    Male          PhD       False   
317        Zhu S       Tian C  Female          PhD        True   
318        Zhu S      Zhang Z  Female          PhD       False   
319        Zhu S        Gao B    Male          PhD       False   

         current job     MD  \
0                 PI  False   
1           Academia  False   
2              Admin  False   
3    Facility leader  False   
4                NaN  False   
..               ...    ...   
315     Senior Staff  False   
316     Senior Staff  False   
317               PI  False   
318         Academia  False   
319         Academia  False   

                                           Affiliation  Country  Ivy league  \
0                                      Yale University      USA        True   
1                              The University of Texas      USA       False   
2                        Tokyo Metropolitan University    Japan       False   
3                Hungarian Academy of Sciences, Szeged  Hungary       False   
4                Hungarian Academy of Sciences, Szeged  Hungary       False   
..                                                 ...      ...         ...   
315                  niversity of Missouri-Kansas City      USA       False   
316  Institute of Zoology, Chinese Academy of Sciences    China       False   
317  Institute of Zoology, Chinese Academy of Sciences    China       False   
318  Institute of Zoology, Chinese Academy of Sciences    China       False   
319  Institute of Zoology, Chinese Academy of Sciences    China       False   

    first_author_key last_author_key  
0            derre i       agaisse h  
1           seong cs     aguilera rj  
2            tsuda m        aigaki t  
3           markus r          ando i  
4              rus f          ando i  
..               ...             ...  
315             ao j           yu xq  
316           yuan y           zhu s  
317           tian c           zhu s  
318          zhang z           zhu s  
319            gao b           zhu s  

[320 rows x 12 columns]

In [ ]:
print(len(authors.article_id.unique()))
authors

400


author_id  article_id               name     sex  leading_author  \
0            2        2053             Brun S    Male           False   
1            3        2053         Lemaitre B    Male            True   
2            1        1650         Charroux B    Male           False   
3           97        2149        Söderhäll K    Male            True   
4            5        2278        Silverman N    Male           False   
..         ...         ...                ...     ...             ...   
792        797        2067         Cerenius L    Male            True   
793        798        2122         Higgins DE    Male            True   
794        799        2184  Mengin-Lecreulx D    Male            True   
795        800        2130           Moore KJ  Female            True   
796        700        1990           Matova N  Female           False   

     first_author career_stage expertise_level         author_key  
0            True     Post-doc     Experienced             brun s  
1           False    Junior PI     Experienced         lemaitre b  
2            True     Post-doc     Experienced         charroux b  
3           False    Senior PI     Experienced        soderhall k  
4            True     Post-doc        Newcomer        silverman n  
..            ...          ...             ...                ...  
792         False    Senior PI     Experienced         cerenius l  
793         False    Senior PI        Newcomer         higgins de  
794         False    Senior PI     Experienced  mengin-lecreulx d  
795         False    Junior PI        Newcomer           moore kj  
796          True     Post-doc        Newcomer           matova n  

[797 rows x 9 columns]

In [53]:
paper_auth_pairs["article_id"] = pd.NA
paper_auth_pairs["article_id"] = paper_auth_pairs["article_id"].astype("Int64")

In [ ]:
for art in authors["article_id"].unique():
    these_authors = authors[authors["article_id"] == art]
    these_first_authors = these_authors[these_authors["first_author"] == True]["author_key"].values
    these_leading_authors = these_authors[these_authors["leading_author"] == True]["author_key"].values
    if (len(these_first_authors) == 1 and len(these_leading_authors) == 1):
        indexes = paper_auth_pairs[(paper_auth_pairs["last_author_key"] == these_leading_authors[0]) & 
                                (paper_auth_pairs["first_author_key"] == these_first_authors[0])].index.to_list()
        if len(indexes) == 1:
            paper_auth_pairs.loc[indexes[0], "article_id"] = art
        elif len(indexes) > 1:
            first_na = pd.isna(paper_auth_pairs.loc[indexes, "article_id"]).idxmax() #idmax returns the index of the first occurrence of the maximum value
            print(these_authors["author_key"].values, pd.isna(paper_auth_pairs.loc[indexes, "article_id"]).values, first_na)
            paper_auth_pairs.loc[first_na, "article_id"] = art
        else:
            print(f"NO MATCH 🛑 {art}, {len(indexes)}, {these_first_authors[0]:<20} -- {these_leading_authors[0]:<20} ... {len(these_authors)}")
    else:
        print(f"SKIPPING {art} because only one={len(these_authors)} author: {these_authors['author_key'].values}")

['georgel p' 'hoffmann ja'] [ True  True] 96
NO MATCH 🛑 2239, 0, hedengren-olcott m   -- taylor bj            ... 2
['georgel p' 'hoffmann ja'] [False  True] 101
NO MATCH 🛑 1947, 0, schneider ds         -- schneider ds         ... 1
NO MATCH 🛑 2730, 0, schneider ds         -- schneider ds         ... 2
NO MATCH 🛑 3447, 0, rizki tm             -- rizki tm             ... 1
NO MATCH 🛑 3448, 0, rizki tm             -- rizki tm             ... 1
NO MATCH 🛑 3292, 0, rizki tm             -- rizki tm             ... 1
NO MATCH 🛑 3273, 0, rizki tm             -- rizki tm             ... 2
['georgel p' 'hoffmann ja'] [False False] 96
NO MATCH 🛑 2764, 0, imler jl             -- imler jl             ... 1
NO MATCH 🛑 3446, 0, nappi aj             -- nappi aj             ... 1


### There is an error above, : 3 hoffman georgel, whereas in the excel there are two...

In [ ]:
paper_auth_pairs[paper_auth_pairs["article_id"].isna()]

last author  first author     Sex PhD Post-doc Become a Pi  \
16       Bellotti RA      Rizki TM    Male           PI        True   
106      Hoffmann JA      Imler JL    Male           PI        True   
240         Rizki TM      RIZKI MT    Male           PI        True   
271    Shahabuddin M   Schneider D    Male           Pi        True   
272  Shirasu-Hiza MM  Schneider DS    Male           Pi        True   
278       Silvers MJ      Nappi AJ    Male           PI        True   
290        Taylor BJ   Hedengren M  Female          PhD       False   

    current job     MD                    Affiliation Country Ivy league  \
16     Deceased  False                      Ann arbor     USA      False   
106          PI  False  CNRS-University of Strasbourg  France      False   
240     Retired  False                      ann arbor     NaN      False   
271          PI   True            Whitehead Institute     USA      False   
272          PI   True   Stanford University Stanford     USA       True   
278          PI  False          Univeristy of Chicago     USA      False   
290         NaN  False        Oregon State University     USA      False   

    first_author_key  last_author_key  article_id  
16          rizki tm      bellotti ra        <NA>  
106         imler jl      hoffmann ja        <NA>  
240         rizki mt         rizki tm        <NA>  
271      schneider d    shahabuddin m        <NA>  
272     schneider ds  shirasu-hiza mm        <NA>  
278         nappi aj       silvers mj        <NA>  
290      hedengren m        taylor bj        <NA>

In [ ]:
paper_auth_pairs.loc[290, "article_id"] = 2239

In [ ]:
paper_auth_pairs[paper_auth_pairs["article_id"].isna()]

last author  first author   Sex PhD Post-doc Become a Pi current job  \
16       Bellotti RA      Rizki TM  Male           PI        True    Deceased   
106      Hoffmann JA      Imler JL  Male           PI        True          PI   
240         Rizki TM      RIZKI MT  Male           PI        True     Retired   
271    Shahabuddin M   Schneider D  Male           Pi        True          PI   
272  Shirasu-Hiza MM  Schneider DS  Male           Pi        True          PI   
278       Silvers MJ      Nappi AJ  Male           PI        True          PI   

        MD                    Affiliation Country Ivy league first_author_key  \
16   False                      Ann arbor     USA      False         rizki tm   
106  False  CNRS-University of Strasbourg  France      False         imler jl   
240  False                      ann arbor     NaN      False         rizki mt   
271   True            Whitehead Institute     USA      False      schneider d   
272   True   Stanford University Stanford     USA       True     schneider ds   
278  False          Univeristy of Chicago     USA      False         nappi aj   

     last_author_key  article_id  
16       bellotti ra        <NA>  
106      hoffmann ja        <NA>  
240         rizki tm        <NA>  
271    shahabuddin m        <NA>  
272  shirasu-hiza mm        <NA>  
278       silvers mj        <NA>

In [ ]:
len(paper_auth_pairs.article_id.unique())

315

### B.2. Read the stat author files, both the sheet for first and last.
- **0: First** `input_data/2025-03-28/2025-03-27_dropbox_6_ReproSci/0_last version/2025_Last version_March 19th/0_First author cleaned_March 18th.xlsx`
  - Pour les statistiques sur ceux qui deviennent PI, Il ne faut pas prendre en compte ceux qui sont surlignés en bleu car ils sont des premiers auteurs qui sont déjà PI. Normalement ce fichier est bon.  
Avec un peu d’approximation, je trouve que sur 291 first authors: 107 become PI et il y  16  ?? (indéterminés)et 13 qui sont déjà PI en bleu :  ce qui fait 107/262=>40,8% (262=291-16-13)  
- **2: Both** `input_data/2025-03-28/2025-03-27_dropbox_6_ReproSci/0_last version/2025_Last version_March 19th/2_2025_March 9th_stats_author.xlsx`
  - obtenu depuis le site
  -  tu trouveras les last avec de nouveau critères  pour historical (tradition), continuty, first and last.  (pour junior et senior il faut voir par articles dans la database)  
  - Feuille 1 : Parmi les first author (1er feuille) :  
    - Zhenting Zhang et Zhaolin Zhang  sont fusionnés en Zhang Z  
    - Zhipeng Wang et Zhi Wang sont fusionné en Wang Z  
    - Hedengren M et Hedengren M et Hedengren-Olcott M ne sont pas fusionnés
  - Feuille 3: pour les last authors:
    - Il faut fusionner Schneider DS et Schneider D -> c'est les mêmes
    - Historical lab (tradition) or not : il faut pendre le colonne Historical lab after 1998 et comparer 1 (historical, trained in hustorical lab) et 0 (non historical) mais ne pas prendre en compte 2.
    - La colonne E first and Last : compte les chercheurs qui apparaissent comme first author  et last authors. Il faudrait comparer 1 (last author qui ont été first author dans une autre labo mais pas PI)  à  0 (last author withouth being first author)  sans prendre en compte les 2 (first author who were already PI when first author).
  

In [ ]:
# read an xlx file in pandas
stat_author_fn = "input_data/2025-03-28/2025-03-27_dropbox_6_ReproSci/0_last version/2025_Last version_March 19th/2_2025_March 9th_stats_author.xlsx"

first_authors_claims = pd.read_excel(stat_author_fn, sheet_name='First')
leading_authors_claims = pd.read_excel(stat_author_fn, sheet_name='Leading')
leading_authors_claims["Authorship"]= "Leading"
first_authors_claims["Authorship"]= "First"

def clean_authors_claims(df):
    df['Sex'] = df['Sex'].map({1: 'Male', 0: 'Female', '??': np.nan, '?': np.nan})
    df = df.drop(columns=[col for col in df.columns if '%' in col])
    df = df.drop(columns=[col for col in df.columns if 'Unnamed' in col])
    df = df.rename(columns={'Conituinity': 'Continuity', 
                                                "Partially verified":"Partially Verified", 
                                                "*Historical lab after 1998":"Historical lab after 1998"
                                                })
    if "Historical lab" in df.columns:
        df['Historical lab'] = df['Historical lab'].astype('boolean')
    if "Continuity" in df.columns:
        df['Continuity'] = df['Continuity'].astype('boolean')
    if "Become a Pi" in df.columns:
        df.loc[df["Become a Pi"] == "??", "Become a Pi"] = np.nan
        df.loc[df["Become a Pi"] == 0, "Become a Pi"] = False
        df.loc[df["Become a Pi"] == 1, "Become a Pi"] = True
    if "MD" in df.columns:
        df.loc[df["MD"] == "??", "MD"] = np.nan
        df.loc[df["MD"] == "?", "MD"] = np.nan
        df.loc[df["MD"] == 'O', "MD"] = False
        df.loc[df["MD"] == 0, "MD"] = False
        df.loc[df["MD"] == 1, "MD"] = True
    if "Ivy league" in df.columns:
        df.loc[df["Ivy league"] == 0, "Ivy league"] = False
        df.loc[df["Ivy league"] == 1, "Ivy league"] = True

    df = df.dropna(subset=['Name'])
    # drop  Schneider DS+D from names as  Schneider DS and  Schneider D are also there in separate rows
    df = df[~(df['Name'] == 'Schneider DS+D')]
    return df


first_authors_claims = clean_authors_claims(first_authors_claims)
leading_authors_claims = clean_authors_claims(leading_authors_claims)

leading_authors_claims = preprocess_utils.build_author_key(leading_authors_claims, author_name_col="Name", key_col="last_author_key")
first_authors_claims = preprocess_utils.build_author_key(first_authors_claims, author_name_col="Name", key_col="first_author_key")

In [5]:
first_authors_claims

Name     Sex  Articles  Major claims  Unchallenged  \
0             Williams MJ    Male         5            14             7   
1              Lemaitre B    Male         4            12             0   
2                   Ha EM  Female         4            10             3   
3               Leulier F    Male         4            10             0   
4           De Gregorio E    Male         3             9             1   
..                    ...     ...       ...           ...           ...   
287           Schlenke TA    Male         1             1             0   
288  Uttenweiler-Joseph S  Female         1             1             0   
289               Wang LN  Female         1             1             0   
290                Yuan Y    Male         1             1             0   
291                Zhou Z    Male         1             1             0   

     Verified  Partially Verified  Mixed  Challenged Authorship  \
0           3                   1      0           3      First   
1          11                   1      0           0      First   
2           1                   2      2           2      First   
3          10                   0      0           0      First   
4           8                   0      0           0      First   
..        ...                 ...    ...         ...        ...   
287         1                   0      0           0      First   
288         1                   0      0           0      First   
289         1                   0      0           0      First   
290         1                   0      0           0      First   
291         1                   0      0           0      First   

         first_author_key  
0             williams mj  
1              lemaitre b  
2                   ha em  
3               leulier f  
4           de gregorio e  
..                    ...  
287           schlenke ta  
288  uttenweiler-joseph s  
289               wang ln  
290                yuan y  
291                zhou z  

[292 rows x 11 columns]

In [8]:
first_authors_claims2_fn = "input_data/2025-03-28/2025-03-27_dropbox_6_ReproSci/0_last version/2025_Last version_March 19th/0_First author cleaned_March 18th.xlsx"
first_authors_claims2 = pd.read_excel(first_authors_claims2_fn)
first_authors_claims2["Name"] = first_authors_claims2["AAA_first author"]
first_authors_claims2["Become a Pi"] = first_authors_claims2["AAA_Become a Pi"]
first_authors_claims2 = first_authors_claims2.drop(columns=["AAA_first author", "AAA_Become a Pi","Source"])

first_authors_claims2 = clean_authors_claims(first_authors_claims2)
print(first_authors_claims2["Sex"].unique(), 
        first_authors_claims2["MD"].unique(), 
        first_authors_claims2["Become a Pi"].unique(), 
        first_authors_claims2["Ivy league"].unique())
first_authors_claims2 = preprocess_utils.build_author_key(first_authors_claims2, author_name_col="Name", key_col="first_author_key")
first_authors_claims2

['Male' 'Female' nan] [False nan True] [True False nan] [False True]


/var/folders/pj/15hxgl0j1wg2w_t5k7bvnq5c0000gn/T/ipykernel_41023/3466221200.py:32: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'False' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.loc[df["Ivy league"] == 0, "Ivy league"] = False


Sex  PhD Post-doc   current job     MD  \
0      Male  Senior Staff            PI  False   
1      Male      Post-doc            PI  False   
2    Female           PhD      Industry  False   
3       NaN           PhD  Senior Staff  False   
4      Male      Post-doc            PI  False   
..      ...           ...           ...    ...   
285  Female           PhD      Academia  False   
286  Female           PhD         Admin  False   
287  Female           PhD            ??  False   
288    Male           PhD            ??  False   
289    Male           PhD            ??  False   

                                           Affiliation  Country Ivy league  \
0                            Forschungszentrum Borstel  Germany      False   
1                               Harvard Medical School      USA       True   
2           University of Massachusetts Medical School      USA      False   
3                    niversity of Missouri-Kansas City      USA      False   
4    Harvard Medical School and Massachusetts Gener...      USA       True   
..                                                 ...      ...        ...   
285  Institute of Zoology, Chinese Academy of Sciences    China      False   
286                             University of Kentucky      USA      False   
287                 University of California San Diego      USA       True   
288                              University of Houston      USA      False   
289                      Shanghai Jiao Tong University    China      False   

              Name Become a Pi first_author_key  
0     Abdelsadik A        True     abdelsadik a  
1        Agaisse H        True        agaisse h  
2       Aggarwal K       False       aggarwal k  
3             Ao J         NaN             ao j  
4    Apidianakis Y        True    apidianakis y  
..             ...         ...              ...  
285        Zhang Z       False          zhang z  
286        Zhang Z       False          zhang z  
287        Zhao HW       False          zhao hw  
288         Zhou Z         NaN           zhou z  
289      Zhuang ZH       False        zhuang zh  

[290 rows x 10 columns]

### B.3 Merge to create a covariate file for Last Author

In [ ]:
paper_auth_pairs_LH = paper_auth_pairs[["last author", "last_author_key", "Affiliation", "Country", "Ivy league"]]
paper_auth_pairs_LH = preprocess_utils.deduplicate_by(paper_auth_pairs_LH, "last author").copy()

In [24]:
 # check for duplicate last_author_key in paper_auth_pairs_LH
paper_auth_pairs_LH[paper_auth_pairs_LH.duplicated('last_author_key', keep=False)]
print(len(paper_auth_pairs_LH[paper_auth_pairs_LH.duplicated('last_author_key', keep=False)]), "duplicates in paper_auth_pairs_LH")
# check for duplicate last_author_key in leading_authors_claims
print(len(leading_authors_claims[leading_authors_claims.duplicated('last_author_key', keep=False)]), "duplicates in leading_authors_claims")

0 duplicates in paper_auth_pairs_LH
0 duplicates in leading_authors_claims


In [25]:
# Perform the outer merge to see what we are missing
all_LH = pd.merge(leading_authors_claims, paper_auth_pairs_LH, on='last_author_key', how='outer')
print(len(leading_authors_claims), len(paper_auth_pairs_LH), len(all_LH))

157 160 161


In [26]:
unique_pairs = all_LH[["Name", "last author", "last_author_key"]].drop_duplicates().sort_values("last_author_key", ascending=True)
print(" ONLY in Pairs               only in Claims        matching key ")
print("-"*70)
for i in range(0, len(unique_pairs)):
    if pd.isna(unique_pairs.iloc[i]['last author']) or pd.isna(unique_pairs.iloc[i]['Name']):
        print('💥 ', end='')
        print(f"{unique_pairs.iloc[i]['last author']:<20} vs  {unique_pairs.iloc[i]['Name']:<20} --> {unique_pairs.iloc[i]['last_author_key']:<20}")
    #else:
    #    print(f"   {unique_pairs.iloc[i]['last author']:<20} vs  {unique_pairs.iloc[i]['Name']:<20} --> {unique_pairs.iloc[i]['last_author_key']:<20}")

 ONLY in Pairs               only in Claims        matching key 
----------------------------------------------------------------------
💥 Bellotti RA          vs  nan                  --> bellotti ra         
💥 nan                  vs  Nappi AJ             --> nappi aj            
💥 Shahabuddin M        vs  nan                  --> shahabuddin m       
💥 Shirasu-Hiza MM      vs  nan                  --> shirasu-hiza mm     
💥 Silvers MJ           vs  nan                  --> silvers mj          


In [27]:
all_LH_inner = pd.merge(leading_authors_claims, paper_auth_pairs_LH, on='last_author_key', how='inner').sort_values("last_author_key", ascending=True)
#all_LH_inner.drop(columns=['last_author_key', 'last author', 'Authorship'], inplace=True)
print(len(all_LH_inner))


assessment_columns = ['Unchallenged', 'Verified', 'Partially Verified', 'Mixed', 'Challenged']
# set type of assessment columns to int
for col in assessment_columns:
    all_LH_inner[col] = all_LH_inner[col].astype(int)
all_LH_inner['Major claims'] = all_LH_inner['Major claims'].astype(int)
all_LH_inner['Articles'] = all_LH_inner['Articles'].astype(int)

for col in assessment_columns:
    all_LH_inner[f'{col}_prop'] = all_LH_inner[col] / all_LH_inner['Major claims']

all_LH_inner = all_LH_inner.apply(preprocess_utils.safe_strip)
all_LH_inner.to_csv('preprocessed_data/LH_inner.csv', index=False)
all_LH_inner

156


Name  Historical lab  Historical lab after 1998  Continuity  \
63     Agaisse H           False                        0.0       False   
107  Aguilera RJ           False                        0.0       False   
64      Aigaki T           False                        0.0       False   
6    Anderson KV           False                        0.0       False   
16        Andó I           False                        0.0        True   
..           ...             ...                        ...         ...   
104         Xu T           False                        0.0       False   
146  Yamaguchi M           False                        0.0       False   
105       Yoo MA           False                        0.0       False   
106        Yu XQ           False                        0.0       False   
47         Zhu S           False                        0.0       False   

     F and L     Sex  Articles  Major claims  Unchallenged  Verified  ...  \
63       1.0    Male         1             3             2         1  ...   
107      0.0    Male         1             2             0         1  ...   
64       0.0    Male         1             3             2         1  ...   
6        0.0  Female         8            20             7        11  ...   
16       0.0    Male         6            12             1         7  ...   
..       ...     ...       ...           ...           ...       ...  ...   
104      0.0    Male         1             3             0         3  ...   
146      0.0    Male         1             2             1         1  ...   
105      0.0  Female         1             3             0         2  ...   
106      0.0    Male         1             3             1         2  ...   
47       0.0    Male         4             6             3         2  ...   

     last_author_key  last author  \
63         agaisse h    Agaisse H   
107      aguilera rj  Aguilera RJ   
64          aigaki t     Aigaki T   
6        anderson kv  Anderson KV   
16            ando i         Ando   
..               ...          ...   
104             xu t         Xu T   
146      yamaguchi m  Yamaguchi M   
105           yoo ma       Yoo MA   
106            yu xq        Yu XQ   
47             zhu s        Zhu S   

                                           Affiliation  Country Ivy league  \
63                                     Yale University      USA       True   
107                            The University of Texas      USA      False   
64                       Tokyo Metropolitan University    Japan      False   
6    Memorial Sloan-Kettering Cancer Center and the...      USA       True   
16               Hungarian Academy of Sciences, Szeged  Hungary      False   
..                                                 ...      ...        ...   
104                 Yale University School of Medicine      USA      False   
146             Aichi Cancer Center Research Institute    Japan      False   
105                          Pusan National University    Korea      False   
106                  niversity of Missouri-Kansas City      USA      False   
47   Institute of Zoology, Chinese Academy of Sciences    China      False   

    Unchallenged_prop Verified_prop Partially Verified_prop Mixed_prop  \
63           0.666667      0.333333                0.000000        0.0   
107          0.000000      0.500000                0.000000        0.0   
64           0.666667      0.333333                0.000000        0.0   
6            0.350000      0.550000                0.000000        0.0   
16           0.083333      0.583333                0.333333        0.0   
..                ...           ...                     ...        ...   
104          0.000000      1.000000                0.000000        0.0   
146          0.500000      0.500000                0.000000        0.0   
105          0.000000      0.666667                0.333333        0.0   
106          0.333333      0.666667                0.000000     

In [28]:
for aut in authors[authors["leading_author"]]["author_key"]:
    if len(all_LH[all_LH["last_author_key"] == aut]) != 1:
        print (f"outer: 💥 {aut}, {len(all_LH[all_LH['last_author_key'] == aut])}")
    if len(all_LH_inner[all_LH_inner["last_author_key"] == aut]) != 1:
        print (f"inner: 💥 {aut}, {len(all_LH_inner[all_LH_inner['last_author_key'] == aut])}")


inner: 💥 nappi aj, 0


### B. First authors

In [ ]:
all_FH = pd.merge(first_authors_claims, first_authors_claims2, on='first_author_key', how='outer', suffixes=('_claims', '_claims2'))
all_LH_inner = pd.merge(first_authors_claims, first_authors_claims2, on='first_author_key', how='inner').sort_values("first_author_key", ascending=True)
print(len(first_authors_claims), len(first_authors_claims2), len(all_FH), len(all_LH_inner))

292 290 295 291


In [18]:
first_authors_claims2.Affiliation.value_counts()

Affiliation
CNRS-University of Strasbourg                  32
University of Stockholm                        29
CNRS_Gif-sur-Yvette_Université Paris-Saclay    20
University of Massachusetts Medical School     15
CNRS University of Marseille                    7
                                               ..
University of Uppsala                           1
Kansas State University                         1
University of Modena and Reggio Emilia          1
University of Glasgow                           1
Shanghai Jiao Tong University                   1
Name: count, Length: 109, dtype: int64

In [29]:
paper_auth_pairs_FH = paper_auth_pairs.drop(columns=['last author']).sort_values("first author", ascending=True)
paper_auth_pairs_FH

first author     Sex  PhD Post-doc Become a Pi   current job  \
243        Abdelsadik A    Male  Senior Staff        True            PI   
220           Agaisse H    Male      Post-doc        True            PI   
92            Agaisse H    Male      Post-doc        True            PI   
274          Aggarwal K  Female           PhD       False      Industry   
315                Ao J     NaN           PhD         NaN  Senior Staff   
..                  ...     ...           ...         ...           ...   
218             Zhang Z  Female           PhD       False         Admin   
87              Zhao HW  Female           PhD       False           NaN   
141              Zhou Z    Male           PhD         NaN           NaN   
79            Zhuang ZH    Male           PhD       False           NaN   
40   de Morais Guedes S  Female           PhD       False           NaN   

        MD                                 Affiliation   Country Ivy league  \
243  False                   Forschungszentrum Borstel   Germany      False   
220  False                      Harvard Medical School       USA       True   
92   False                      Harvard Medical School       USA       True   
274  False  University of Massachusetts Medical School       USA      False   
315  False           niversity of Missouri-Kansas City       USA      False   
..     ...                                         ...       ...        ...   
218  False                      University of Kentucky       USA      False   
87   False          University of California San Diego       USA       True   
141  False                       University of Houston       USA      False   
79   False               Shanghai Jiao Tong University     China      False   
40   False                        University of Aveiro  Portugal      False   

       first_author_key last_author_key  
243        abdelsadik a        roeder t  
220           agaisse h      perrimon n  
92            agaisse h      higgins de  
274          aggarwal k     silverman n  
315                ao j           yu xq  
..                  ...             ...  
218             zhang z        palli sr  
87              zhao hw       haddad gg  
141              zhou z     kimbrell da  
79            zhuang zh           ge bx  
40   de morais guedes s     domingues p  

[320 rows x 11 columns]

In [30]:
paper_auth_pairs_FH = preprocess_utils.deduplicate_by(paper_auth_pairs_FH, "first author")

            among ['perrimon n', 'higgins de'].
            among ['Post-doc', 'PhD'].
            among ['Harvard Medical School and Massachusetts General Hospital', 'Massachusetts General Hospital'].
            among ['rahme lg', 'mylonakis e'].
            among ['PhD', 'Post-doc'].
            among ['CEA Grenoble', 'Université de Toulouse'].
            among ['fauvarque mo', 'waltzer l'].
            among ['CNRS Gif-sur-Yvette Université Paris-Saclay', 'CNRS_Gif-sur-Yvette_Université Paris-Saclay'].
            among ['boccard f', 'lemaitre b'].
            among ['CNRS_Gif-sur-Yvette_Université Paris-Saclay', 'CNRS Gif-sur-Yvette Université Paris-Saclay'].
            among ['lemaitre b', 'brey pt'].
            among ['lagueux m', 'hoffmann ja'].
            among ['Massachusetts General Hospital', 'Harvard Medical School,Boston'].
            among ['white k', 'ezekowitz ra'].
            among ['Univeristy of Freiburg', 'Johannes Gutenberg Universität'].
            among [

In [ ]:

# check for duplicate first_author_key in paper_auth_pairs_LH
print(len(paper_auth_pairs_FH[paper_auth_pairs_FH.duplicated('first_author_key', keep=False)]), "duplicates in paper_auth_pairs_LH")
# check for duplicate first_author_key in leading_authors_claims
print(len(first_authors_claims[first_authors_claims.duplicated('first_author_key', keep=False)]), "duplicates in leading_authors_claims")


all_FH = pd.merge(first_authors_claims, paper_auth_pairs_FH, on='first_author_key', how='outer', suffixes=('', '_pairs'))
print(len(first_authors_claims), len(paper_auth_pairs_FH), len(all_FH))

print(" ONLY in Pairs               only in Claims        matching key ")
print("-"*70)
unique_pairs = all_FH[["Name", "first author", "first_author_key"]].drop_duplicates().sort_values("first_author_key", ascending=True)
for i in range(0, len(unique_pairs)):
    if pd.isna(unique_pairs.iloc[i]['first author']) or pd.isna(unique_pairs.iloc[i]['Name']):
        print('💥 ', end='')
        print(f" {unique_pairs.iloc[i]['first author']:<20}  {unique_pairs.iloc[i]['Name']:<20} --> {unique_pairs.iloc[i]['first_author_key']:<20}")

0 duplicates in paper_auth_pairs_LH
4 duplicates in leading_authors_claims
292 291 294
 ONLY in Pairs               only in Claims        matching key 
----------------------------------------------------------------------
💥  nan                   Hedengren-Olcott M   --> hedengren-olcott m  
💥  RIZKI MT              nan                  --> rizki mt            
💥  Schneider D           nan                  --> schneider d         


In [32]:


all_FH = all_FH.apply(preprocess_utils.safe_strip)

In [ ]:
all_FH_inner = pd.merge(first_authors_claims, paper_auth_pairs_FH, on='first_author_key', how='inner', suffixes=('', '_pairs')).sort_values("first_author_key", ascending=True)
for idx, row in all_FH_inner.iterrows():
    if row["Sex"] != row["Sex_pairs"]:
        print(f"confligcthing sex for {row['Name']} : {row['Sex']} vs  {row['Sex_pairs']}")
all_FH_inner.drop(columns=["Sex_pairs"], inplace=True)
print(len(all_FH_inner))
all_FH_inner = all_FH_inner.apply(preprocess_utils.safe_strip)

assessment_columns = ['Unchallenged', 'Verified', 'Partially Verified', 'Mixed', 'Challenged']
# set type of assessment columns to int
for col in assessment_columns:
    all_FH_inner[col] = all_FH_inner[col].astype(int)
all_FH_inner['Major claims'] = all_FH_inner['Major claims'].astype(int)
all_FH_inner['Articles'] = all_FH_inner['Articles'].astype(int)

for col in assessment_columns:
    all_FH_inner[f'{col}_prop'] = all_FH_inner[col] / all_FH_inner['Major claims']

all_FH_inner.to_csv('preprocessed_data/FH_inner.csv', index=False)
all_FH_inner

confligcthing sex for Ao J : Male vs  nan
confligcthing sex for Avila A : Male vs  Female
confligcthing sex for Cheng W : Female vs  nan
confligcthing sex for Chiu H : Male vs  Female
confligcthing sex for Daffre S : Male vs  Female
confligcthing sex for Ertürk-Hasdemir D : Male vs  Female
confligcthing sex for Han SH : Male vs  Female
confligcthing sex for Hedengren M : Male vs  Female
confligcthing sex for Kim YS : Female vs  Male
confligcthing sex for Kwon EJ : Male vs  Female
confligcthing sex for Lindmark H : Female vs  Male
confligcthing sex for Luce-Fedrow A : Female vs  Male
confligcthing sex for Michel T : Female vs  Male
confligcthing sex for Mishima Y : Female vs  Male
confligcthing sex for Nam HJ : Male vs  Female
confligcthing sex for Peng J : Male vs  Female
confligcthing sex for Persson C : Male vs  Female
confligcthing sex for Pham LN : Male vs  Female
confligcthing sex for Ren C : Female vs  Male
confligcthing sex for Schmidt RL : Female vs  Male
confligcthing sex for 

Name     Sex  Articles  Major claims  Unchallenged  Verified  \
79    Abdelsadik A    Male         1             3             3         0   
19       Agaisse H    Male         2             6             2         4   
195     Aggarwal K  Female         1             2             0         2   
80            Ao J    Male         1             3             1         2   
20   Apidianakis Y    Male         3             6             4         2   
..             ...     ...       ...           ...           ...       ...   
193  Zettervall CJ    Male         1             3             0         3   
273        Zhang Z  Female         2             2             1         1   
194        Zhao HW  Female         1             3             2         1   
290         Zhou Z    Male         1             1             0         1   
42       Zhuang ZH    Male         2             6             1         4   

     Partially Verified  Mixed  Challenged Authorship  ...     MD  \
79                    0      0           0      First  ...  False   
19                    0      0           0      First  ...  False   
195                   0      0           0      First  ...  False   
80                    0      0           0      First  ...  False   
20                    0      0           0      First  ...  False   
..                  ...    ...         ...        ...  ...    ...   
193                   0      0           0      First  ...  False   
273                   0      0           0      First  ...  False   
194                   0      0           0      First  ...  False   
290                   0      0           0      First  ...  False   
42                    1      0           0      First  ...  False   

                                           Affiliation  Country Ivy league  \
79                           Forschungszentrum Borstel  Germany      False   
19                              Harvard Medical School      USA       True   
195         University of Massachusetts Medical School      USA      False   
80                   niversity of Missouri-Kansas City      USA      False   
20   Harvard Medical School and Massachusetts Gener...      USA       True   
..                                                 ...      ...        ...   
193                            University of Stockholm   Sweden      False   
273  Institute of Zoology, Chinese Academy of Sciences    China      False   
194                 University of California San Diego      USA       True   
290                              University of Houston      USA      False   
42                       Shanghai Jiao Tong University    China      False   

    last_author_key Unchallenged_prop Verified_prop Partially Verified_prop  \
79         roeder t          1.000000      0.000000                0.000000   
19       perrimon n          0.333333      0.666667                0.000000   
195     silverman n          0.000000      1.000000                0.000000   
80            yu xq          0.333333      0.666667                0.000000   
20         rahme lg          0.666667      0.333333                0.000000   
..              ...               ...           ...                     ...   
193      hultmark d          0.000000      1.000000                0.000000   
273           zhu s          0.500000      0.500000                0.000000   
194       haddad gg          0.666667      0.333333                0.000000   
290     kimbrell da          0.000000      1.000000                0.000000   
42            ge bx          0.166667      0.666667                0.166667   

     Mixed_prop Challenged_prop  
79          0.0             0.0  
19          0.0             0.0  
195         0.0             0.0  
80          0.0             0.0  
20          0.0             0.0  
..          ...             ...  
193         0.0             0.0  
273         0.0             0.0  
194         0.0             0.0  
290         0.0       

In [34]:
for aut in authors[authors["first_author"]]["author_key"]:
    #if len(all_FH[all_LH["first_author_key"] == aut]) != 1:
    #    print (f"outer: 💥 {aut}, {len(all_FH[all_FH['author_key'] == aut])}")
    if len(all_FH_inner[all_FH_inner["first_author_key"] == aut]) != 1:
        print (f"inner: 💥 {aut}, {len(all_FH_inner[all_FH_inner['first_author_key'] == aut])}")

inner: 💥 hedengren-olcott m, 0
inner: 💥 hedengren m, 2
inner: 💥 hedengren m, 2
inner: 💥 kim ys, 2
inner: 💥 kim ys, 2
inner: 💥 kim ys, 2


In [35]:
all_FH_inner.columns

Index(['Name', 'Sex', 'Articles', 'Major claims', 'Unchallenged', 'Verified',
       'Partially Verified', 'Mixed', 'Challenged', 'Authorship',
       'first_author_key', 'first author', 'PhD Post-doc', 'Become a Pi',
       'current job', 'MD', 'Affiliation', 'Country', 'Ivy league',
       'last_author_key', 'Unchallenged_prop', 'Verified_prop',
       'Partially Verified_prop', 'Mixed_prop', 'Challenged_prop'],
      dtype='object')

In [ ]:
for col in all_FH_inner.columns:
    print(70*"-")
    print(all_FH_inner[col].value_counts())

## Summary
So how to analyze ? The key works well to merge at least, so let's have it. I think we should not do by authors because the aurhors are fleeting. Whether it's first or not. Let's do it by claim when it is needed. I think it's much more correct.